src/hostapi/pulseaudio/pa_linux_pulseaudio.c

```diff
@@ -0,0 +1,1285 @@
+        default:
+            PA_DEBUG( ("Portaudio %s: FAILED '%s'\n",
+                      __FUNCTION__,
+                      pa_strerror( pa_context_errno( pa_stream_get_context( s ) ) ) ) );
+
+            break;
+    }
+}
+
+void PaPulseAudio_StreamUnderflowCb( pa_stream * s,
+                                     void *userdata )
+{
+    PaPulseAudio_Stream *stream = (PaPulseAudio_Stream *) userdata;
+
+
+    /* If this is null we have big problems and we probably are out of memory */
+    if( !s )
+    {
+        PA_PULSEAUDIO_SET_LAST_HOST_ERROR( 0,
+                                           "PaPulseAudio_StreamUnderflowCb: Out of memory" );
+        return;
+    }
+
+    stream->underflows++;
+
+    if( stream->underflows >= 6 && stream->latency < 2000000 )
+    {
+        stream->latency = (stream->latency * 3) / 2;
+        stream->bufferAttr.maxlength =
+            pa_usec_to_bytes(stream->latency, &stream->outSampleSpec);
+        stream->bufferAttr.tlength =
+            pa_usec_to_bytes(stream->latency, &stream->outSampleSpec);
+        pa_stream_set_buffer_attr(s, &stream->bufferAttr, NULL, NULL);
+        stream->underflows = 0;
+        PA_DEBUG(("Portaudio %s: latency increased to %d\n", __FUNCTION__,
+                  stream->latency));
+    }
+
+    pa_threaded_mainloop_signal( stream->mainloop,
+                                 0 );
+}
+
+
+PaError PaPulseAudio_Initialize( PaUtilHostApiRepresentation ** hostApi,
+                                 PaHostApiIndex hostApiIndex )
+{
+    PaError result = paNoError;
+    int i;
+    int deviceCount;
+    int l_iRtn = 0;
+    PaPulseAudio_HostApiRepresentation *l_ptrPulseAudioHostApi = NULL;
+    PaDeviceInfo *deviceInfoArray = NULL;
+
+    pa_operation *l_ptrOperation = NULL;
+
+    l_ptrPulseAudioHostApi = PaPulseAudio_New();
+
+    if( !l_ptrPulseAudioHostApi )
+    {
+        result = paInsufficientMemory;
+        goto error;
+    }
+
+    l_ptrPulseAudioHostApi->allocations = PaUtil_CreateAllocationGroup();
+
+    if( !l_ptrPulseAudioHostApi->allocations )
+    {
+        result = paInsufficientMemory;
+        goto error;
+    }
+
+    l_ptrPulseAudioHostApi->hostApiIndex = hostApiIndex;
+    *hostApi = &l_ptrPulseAudioHostApi->inheritedHostApiRep;
+    (*hostApi)->info.structVersion = 1;
+    (*hostApi)->info.type = paPulseAudio;
+    (*hostApi)->info.name = "PulseAudio";
+
+    (*hostApi)->info.defaultInputDevice = paNoDevice;
+    (*hostApi)->info.defaultOutputDevice = paNoDevice;
+
+    /* Connect to server */
+    pa_threaded_mainloop_lock( l_ptrPulseAudioHostApi->mainloop );
+    l_iRtn = pa_context_connect( l_ptrPulseAudioHostApi->context,
+                                 NULL,
+                                 0,
+                                 NULL );
+
+    if( l_iRtn < 0 )
+    {
+        PA_PULSEAUDIO_SET_LAST_HOST_ERROR( 0,
+                                           "PulseAudio_Initialize: Can't connect to server");
+        result = paNoError;
```

Comment:
Is paNoError correct here?

Reply:
Definetly not. I have to look better alternative